From 9e3ad206c881956d2adf7b8f65bb6b02e56e8d88 Mon Sep 17 00:00:00 2001
From: Jiacheng Zheng <jiaczhen@codeaurora.org>
Date: Thu, 2 Jul 2020 16:27:12 +0800
Subject: [PATCH] ion use system heap type instead of carve out type

Carve out ion allocates physical continuous memory and will
fragment under heavy concurrency and finally causes memory
allocation failure. System heap ion does not guarantee
physical continuous memory but guarantees memory allocation
success as long as there is enough memory in the system.

Change-Id: Idfac47793536fb071e32dfbb11a09e80a8a2c81f
Signed-off-by: Jiacheng Zheng <jiaczhen@codeaurora.org>
---
 drivers/staging/android/ion/ion_hvenv_driver.c | 4 +---
 drivers/staging/android/ion/ion_system_heap.c  | 9 +--------
 2 files changed, 2 insertions(+), 11 deletions(-)

diff --git a/drivers/staging/android/ion/ion_hvenv_driver.c b/drivers/staging/android/ion/ion_hvenv_driver.c
index 5b9870d..7e463ae 100644
--- a/drivers/staging/android/ion/ion_hvenv_driver.c
+++ b/drivers/staging/android/ion/ion_hvenv_driver.c
@@ -39,10 +39,8 @@ static struct ion_heap **heaps;
 static struct ion_platform_heap dummy_heaps[] = {
 	{
 		.id	= ION_SYSTEM_HEAP_ID,
-		.type	= ION_HEAP_TYPE_CARVEOUT,
+		.type	= ION_HEAP_TYPE_SYSTEM,
 		.name	= "ion_system_heap",
-		.base	= 0xF0000000,
-		.size	= 0x10000000,
 	},
 	{
 		.id	= ION_AUDIO_HEAP_ID,
diff --git a/drivers/staging/android/ion/ion_system_heap.c b/drivers/staging/android/ion/ion_system_heap.c
index d2bc7e5..bd25fae 100644
--- a/drivers/staging/android/ion/ion_system_heap.c
+++ b/drivers/staging/android/ion/ion_system_heap.c
@@ -432,9 +432,7 @@ static int ion_system_heap_allocate(struct ion_heap *heap,
 		max_order = info->order;
 		i++;
 	}
-#ifndef CONFIG_MSM_GVM_QUIN
 	ret = msm_ion_heap_alloc_pages_mem(&data);
-#endif
 
 	if (ret)
 		goto err;
@@ -486,9 +484,7 @@ static int ion_system_heap_allocate(struct ion_heap *heap,
 
 	} while (sg);
 
-#ifndef CONFIG_MSM_GVM_QUIN
 	ret = msm_ion_heap_pages_zero(data.pages, data.size >> PAGE_SHIFT);
-#endif
 
 	if (ret) {
 		pr_err("Unable to zero pages\n");
@@ -509,9 +505,8 @@ static int ion_system_heap_allocate(struct ion_heap *heap,
 	buffer->priv_virt = table;
 	if (nents_sync)
 		sg_free_table(&table_sync);
-#ifndef CONFIG_MSM_GVM_QUIN
 	msm_ion_heap_free_pages_mem(&data);
-#endif
+
 
 	return 0;
 
@@ -532,9 +527,7 @@ err_free_sg:
 err1:
 	kfree(table);
 err_free_data_pages:
-#ifndef CONFIG_MSM_GVM_QUIN
 	msm_ion_heap_free_pages_mem(&data);
-#endif
 err:
 	list_for_each_entry_safe(info, tmp_info, &pages, list) {
 		free_buffer_page(sys_heap, buffer, info->page, info->order);
-- 
2.7.4

